tools/server/server-context.h

```diff
@@ -0,0 +1,276 @@
+#include "server-common.h"
+#include "server-http.h"
+#include "server-task.h"
+#include "server-queue.h"
+
+#include "arg.h"
+#include "common.h"
+#include "llama.h"
+#include "log.h"
+#include "sampling.h"
+#include "speculative.h"
+#include "mtmd.h"
+#include "mtmd-helper.h"
+
+#include <atomic>
+#include <cstddef>
+#include <cinttypes>
+#include <memory>
+#include <signal.h>
+#include <thread>
+#include <unordered_set>
+
+// fix problem with std::min and std::max
+#if defined(_WIN32)
+#define WIN32_LEAN_AND_MEAN
+#ifndef NOMINMAX
+#   define NOMINMAX
+#endif
+#include <windows.h>
+#endif
+
+using json = nlohmann::ordered_json;
+
+constexpr int HTTP_POLLING_SECONDS = 1;
+
+// state diagram: https://github.com/ggml-org/llama.cpp/pull/9283
+enum slot_state {
+    SLOT_STATE_IDLE,
+    SLOT_STATE_STARTED, // TODO: this state is only used for setting up the initial prompt processing; maybe merge it with launch_slot_with_task in the future
+    SLOT_STATE_PROCESSING_PROMPT,
+    SLOT_STATE_DONE_PROMPT,
+    SLOT_STATE_GENERATING,
+};
+
+enum server_state {
+    SERVER_STATE_LOADING_MODEL,  // Server is starting up, model not fully loaded yet
+    SERVER_STATE_READY,          // Server is ready and model is loaded
+};
+
+// forward declarations
+struct server_slot;
+
+// proxy for std::vector to allow forward declaration of server_slot
+struct server_slots_t {
+    ~server_slots_t();
+    std::vector<server_slot*> data;
+    size_t size() const { return data.size(); }
+    server_slot & operator[](size_t idx) { return *(data[idx]); }
+    server_slot & operator[](size_t idx) const { return *(data[idx]); }
+    void clear();
+    server_slot & create();
+    struct iterator {
+        typename std::vector<server_slot*>::iterator it;
+        iterator(typename std::vector<server_slot*>::iterator i) : it(i) {}
+        server_slot & operator*() { return **it; }
+        iterator & operator++() { ++it; return *this; }
+        bool operator!=(const iterator& other) const { return it != other.it; }
+    };
+    iterator begin() { return iterator(data.begin()); }
+    iterator end() { return iterator(data.end()); }
+};
```

Comment:
Can't wrap my head around how to forward declare `server_slot` while using it inside a `std::vector`, so I came up with this solution.

Pimpl can also be an alternative, though I doubt if it requires quite a lot of changes in the code (I'm preventing changes as much as possible, to avoid merge conflicts with other PRs). WDYT? @ggerganov 

---

Edit: if we decide to use pimpl, then I think most of the code will be moved into the pimpl and the main class will only expose these 4 function: init(), load_model(), start_loop(), terminate() ; It will be a big migration though

Reply:
Implemented in https://github.com/ggml-org/llama.cpp/pull/17595/commits/9a7b4f36c851526a93bc36e846351f287ee51452 : 
- `server_context` is moved to `server_context_impl` (just renaming struct, the code is left untouched)
- the new `server_context` only expose: init(), load_model(), start_loop(), terminate()

@ggerganov Would be nice if you can review it in a short time frame (ideally before https://github.com/ggml-org/llama.cpp/pull/17470 as some conflicts are inevitable). Thanks!